card/src/test/java/com/adyen/checkout/card/DefaultCardDelegateTest.kt

```diff
@@ -0,0 +1,804 @@
+import com.adyen.checkout.components.ui.FieldState
+import com.adyen.checkout.components.ui.Validation
+import com.adyen.checkout.components.util.PaymentMethodTypes
+import com.adyen.checkout.core.api.Environment
+import com.adyen.checkout.cse.CardEncrypter
+import com.adyen.checkout.cse.GenericEncrypter
+import com.adyen.checkout.cse.test.TestCardEncrypter
+import com.adyen.checkout.cse.test.TestGenericEncrypter
+import com.adyen.checkout.test.TestDispatcherExtension
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.test.UnconfinedTestDispatcher
+import kotlinx.coroutines.test.runTest
+import org.junit.jupiter.api.Assertions.assertEquals
+import org.junit.jupiter.api.Assertions.assertFalse
+import org.junit.jupiter.api.Assertions.assertNull
+import org.junit.jupiter.api.Assertions.assertTrue
+import org.junit.jupiter.api.BeforeEach
+import org.junit.jupiter.api.DisplayName
+import org.junit.jupiter.api.Nested
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.extension.ExtendWith
+import org.mockito.junit.jupiter.MockitoExtension
+import java.util.Locale
+
+@OptIn(ExperimentalCoroutinesApi::class)
+@ExtendWith(MockitoExtension::class, TestDispatcherExtension::class)
+internal class DefaultCardDelegateTest {
+
+    private lateinit var cardEncrypter: TestCardEncrypter
+    private lateinit var genericEncrypter: TestGenericEncrypter
+    private lateinit var publicKeyRepository: TestPublicKeyRepository
+    private lateinit var addressRepository: TestAddressRepository
+    private lateinit var detectCardTypeRepository: TestDetectCardTypeRepository
+    private lateinit var delegate: DefaultCardDelegate
+
+    @BeforeEach
+    fun before() {
+        cardEncrypter = TestCardEncrypter()
+        genericEncrypter = TestGenericEncrypter()
+        publicKeyRepository = TestPublicKeyRepository()
+        addressRepository = TestAddressRepository()
+        detectCardTypeRepository = TestDetectCardTypeRepository()
+        delegate = createCardDelegate()
+    }
+
+    @Test
+    fun `when component is not initialized, then an error is propagated`() = runTest {
+        publicKeyRepository.shouldReturnError = true
+
+        delegate.exceptionFlow.test {
+            delegate.initialize(CoroutineScope(UnconfinedTestDispatcher()))
+
+            val exception = expectMostRecentItem()
+
+            assertEquals(publicKeyRepository.errorResult.exceptionOrNull(), exception.cause)
+
+            cancelAndIgnoreRemainingEvents()
+        }
+    }
+
+    @Nested
+    @DisplayName("when input data changes and")
+    inner class InputDataChangedTest {
+        @Test
+        fun `address configuration is none, then countries and states should not be fetched`() = runTest {
+            addressRepository.countriesFlow.test {
+                addressRepository.statesFlow.test {
+                    delegate.initialize(CoroutineScope(UnconfinedTestDispatcher()))
```

Comment:
It would be better to pass the `TestScope` here to make sure all coroutines are correctly canceled when the test finishes. So you can do `delegate.initialize(this@runTest)`

Reply:
This doesn't work :/